### ISSUE TITLE
Increase allocation cap from mempool from arbitary limit to INT_MAX

### ISSUE THREAD

**Issue by pplimport** [UIUC-PPL/charm#84]
*Original author: Celso Mendes*
*Original issue: https://charm.cs.illinois.edu/redmine/issues/84*

---
There seems to be a serious problem in the memory allocation scheme provided by mempool under the gemini_gni-crayxe layer for AMPI. A simple example shown here demonstrates the problem, which is arising in actual executions of Jacobi and BRAMS, under gemini_gni-crayxe on Blue Waters and JYC with GNU compilers.

Although the mempool source code (mempool.c, mempool.h) looks to have been designed to support up to 1GB of allocation, the actual limit available is much lower. The code example below, which declares a 1 GB array that is sliced across the number of processors in use, will work fine with 8 processors (i.e. 128 MB allocated by each processor), but when it's run with 4 processors (i.e. 256 MB per processor) it fails on every processor with error-messages like

[0] CmiAbort: Mempool-requested slot is more than what mempool can provide as    one chunk, increase cutOffNum and cutoffPoints in mempool

So, apparently there's something that is wrongly limiting the amount of memory that one can allocate, which would mean a bug. Also, 1GB is probably a low limit these days, anyway, as a Cray-XE node has 64 GB of memory, and even when running one process per integer core, each process has 2GB of memory available -- in fact, using fewer tasks per node, each process can use even more than that. So, once the current bug is fixed, it would also be good to increase that current 1 GB limit, if at all possible.

This test was done with GNU compilers (module PrgEnv-gnu loaded) on JYC, using AMPI from a Charm checkout on March-6/2013. I left the example code shown below, plus a Makefile, job-submission script and execution outputs, on JYC under ~cmendes/BUG-AMPI/BUG-3/

AMPI Code Example:
```
SUBROUTINE MPI_MAIN
include 'mpif.h'

PARAMETER (N=16384)
REAL, ALLOCATABLE :: A(:,:)
INTEGER comm, myrank, p, ierr, m

comm=MPI_COMM_WORLD
CALL MPI_INIT(IERR)
CALL MPI_COMM_SIZE(comm, p, ierr)
CALL MPI_COMM_RANK(comm, myrank, ierr)

m = N/p
ALLOCATE (A(N,m))

CALL MPI_BARRIER(comm, ierr)
print *,"[",myrank,"] Done!"
CALL MPI_FINALIZE(IERR)

END

```

**Comment by PhilMiller** [comment]
*Original date: 2013-03-11 17:30:51*

---
Format code as code, and remove signature, since the tracker records issue authors explicitly.

**Comment by PhilMiller** [comment]
*Original date: 2013-05-30 19:14:00*

---
If I recall correctly from yesterday's discussion, this is an issue of isomalloc anywhere we use it, not specifically Cray uGNI builds. The fix for it entails the `int -> size_t` conversion, making it too intrusive for a stable release. Nikhil, please confirm and if I've got that right, change title and target version accordingly.

**Comment by PhilMiller** [comment]
*Original date: 2013-06-03 20:28:18*

---
Just make the much narrower change based on the charm-6.5 stable branch.

**Comment by PhilMiller** [comment]
*Original date: 2013-06-04 05:45:34*

---
I'm upping priority on this for visibility in searches for issues set higher than 'Normal' - please don't take this as me being a nag.

**Comment by nikhil-jain** [comment]
*Original date: 2013-06-07 01:44:29*

---
Mempool has been modified in branch isomalloc_pup. However this change entails moving away from use of int to size for various variables types, notably in PUP. Testing needs to be done.

**Comment by PhilMiller** [comment]
*Original date: 2013-06-07 03:04:20*

---
Apparently, I did things backwards and caused some confusion. I left this one assigned to 6.5.1 to track the simple fix you made. The other bug was already targeted at 6.6, with the intention being to track the invasive change there. We can close each when we've got testing reports from Celso and/or Alvaro that the fix works appropriately.

**Comment by nikhil-jain** [comment]
*Original date: 2013-06-07 03:15:14*

---
It was my misunderstanding. This issue is indeed for the temporary fix, and has been committed to 6.5 branch.
